include/oqpi/synchronization/posix/posix_event.hpp

```diff
@@ -0,0 +1,242 @@
+    template<typename _ResetPolicy>
+    using event_impl = class posix_event<_ResetPolicy>;
+    //----------------------------------------------------------------------------------------------
+
+
+    //----------------------------------------------------------------------------------------------
+    template<typename _ResetPolicy>
+    class posix_event
+            : protected _ResetPolicy
+    {
+    protected:
+        //------------------------------------------------------------------------------------------
+        using native_handle_type = sem_t*;
+
+    protected:
+        //------------------------------------------------------------------------------------------
+        posix_event(const std::string &name, sync_object_creation_options creationOption)
+                : handle_(nullptr)
+                , isLocalSyncObject_(name.empty())
+        {
+            if(isLocalSyncObject_)
+            {
+                // Create an unnamed semaphore.
+                handle_ = new sem_t;
+
+                const auto error = sem_init(handle_, 0, 0);
+                if(error == -1)
+                {
+                    oqpi_error("sem_init failed with error %d", errno);
+                }
+            }
+            else
+            {
+                // Create a named semaphore.
+                const auto validName = isNameValid(name);
+                oqpi_check(validName);
+
+                if(!validName)
+                {
+                    oqpi_error("the name \"%s\" you provided is not valid for a posix semaphore.", name.c_str());
+                    return;
+                }
+
+                // If a semaphore with name doesn't exist, then it is created.
+                handle_ = sem_open(name.c_str(), O_CREAT, S_IRUSR | S_IWUSR, 0);
+
+                if(handle_ == SEM_FAILED)
+                {
+                    oqpi_error("sem_open failed with error %d", errno);
+                }
+            }
+        }
+
+        //------------------------------------------------------------------------------------------
+        posix_event(posix_event &&rhs)
+                : handle_(rhs.handle_)
+                , isLocalSyncObject_(rhs.isLocalSyncObject_)
+        {
+            rhs.handle_ = nullptr;
+        }
+
+        //------------------------------------------------------------------------------------------
+        ~posix_event()
+        {
+            if (handle_)
+            {
+                if(isLocalSyncObject_)
+                {
+                    sem_destroy(handle_);
+                    delete handle_;
+                }
+                else
+                {
+                    sem_close(handle_);
+                }
+                handle_ = nullptr;
+            }
+        }
+
+        //------------------------------------------------------------------------------------------
+        posix_event& operator =(posix_event &&rhs)
+        {
+            if (this != &rhs)
```

Comment:
We need to check that the moved to object (this) is invalid, otherwise it's a potential memleak

Reply:
if it is still valid, do we want to call its destructor and then complete the move?